File: services/2-redis-cache/go/internal/cache/cache.go
// Пакет cache - обёртка над Redis для итерации 2 (§2.5).
//
// Стратегия - ленивая загрузка (cache-aside):
//   - Get возвращает закэшированное значение либо ошибку redis.Nil
//     при промахе;
//   - Set сохраняет значение с заданным TTL.
//
// Сериализация - JSON (тот же формат, в котором сервис отдаёт ответ
// клиенту), чтобы попадание в кэш можно было отдать без пересборки.
package cache

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
	"github.com/redis/go-redis/v9"
)

var (
	cacheHits = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "rec_cache_hits_total",
		Help: "Попадания в кэш",
	}, []string{"kind"})
	cacheMisses = promauto.NewCounterVec(prometheus.CounterOpts{
		Name: "rec_cache_misses_total",
		Help: "Промахи кэша",
	}, []string{"kind"})
)

type Cache struct {
	client *redis.Client
	ttl    time.Duration
}

// New создаёт клиента Redis по DSN-строке (redis://host:port/db).
// Делает Ping для проверки соединения на старте.
func New(ctx context.Context, redisURL string, ttlSeconds int) (*Cache, error) {
	opts, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, fmt.Errorf("разбор адреса Redis: %w", err)
	}
	client := redis.NewClient(opts)
	if err := client.Ping(ctx).Err(); err != nil {
		return nil, fmt.Errorf("проверка доступности Redis: %w", err)
	}
	return &Cache{
		client: client,
		ttl:    time.Duration(ttlSeconds) * time.Second,
	}, nil
}

func (c *Cache) Close() error {
	return c.client.Close()
}

func key(prefix string, userID, n int) string {
	return fmt.Sprintf("%s:%d:n=%d", prefix, userID, n)
}

// Get десериализует JSON-значение в out. Возвращает true при попадании
// в кэш, false при промахе (включая ошибку redis.Nil). Любая другая
// ошибка возвращается явно.
func (c *Cache) Get(ctx context.Context, prefix string, userID, n int, out any) (bool, error) {
	raw, err := c.client.Get(ctx, key(prefix, userID, n)).Bytes()
	if err == redis.Nil {
		cacheMisses.WithLabelValues(prefix).Inc()
		return false, nil
	}
	if err != nil {
		return false, err
	}
	if err := json.Unmarshal(raw, out); err != nil {
		return false, err
	}
	cacheHits.WithLabelValues(prefix).Inc()
	return true, nil
}

// Set сериализует value в JSON и сохраняет с TTL.
func (c *Cache) Set(ctx context.Context, prefix string, userID, n int, value any) error {
	raw, err := json.Marshal(value)
	if err != nil {
		return err
	}
	return c.client.Set(ctx, key(prefix, userID, n), raw, c.ttl).Err()
}

File: services/2-redis-cache/python/app/cache.py
"""
Кэш рекомендаций в Redis (итерация 2, §2.5).

Стратегия - ленивая загрузка (cache-aside): при попадании в кэш
возвращается сериализованный результат; при промахе - алгоритм
вычисляет результат, сохраняет его в кэш с TTL и возвращает.

Ключи:
    recommendations:{user_id}:n={n}  - результат /recommendations/{user_id}?n=...
    popular:{user_id}:n={n}          - результат /recommendations/{user_id}/popular?n=...

TTL - единый, задаётся CACHE_TTL_SECONDS (по умолчанию 60 секунд).
"""
from __future__ import annotations

import json

import redis.asyncio as redis
from prometheus_client import Counter

from app.config import REDIS_URL, CACHE_TTL_SECONDS

CACHE_HITS = Counter("rec_cache_hits_total", "Попадания в кэш", ["kind"])
CACHE_MISSES = Counter("rec_cache_misses_total", "Промахи кэша", ["kind"])

_pool: redis.Redis | None = None


async def init_pool() -> None:
    """Инициализация пула соединений с Redis."""
    global _pool
    _pool = redis.from_url(REDIS_URL, encoding="utf-8", decode_responses=True)
    # ping проверяет, что Redis доступен на старте
    await _pool.ping()


async def close_pool() -> None:
    global _pool
    if _pool is not None:
        await _pool.close()
        _pool = None


def _key(prefix: str, user_id: int, n: int) -> str:
    return f"{prefix}:{user_id}:n={n}"


async def get(prefix: str, user_id: int, n: int) -> list[dict] | None:
    """Возвращает закэшированный результат либо None при промахе."""
    if _pool is None:
        return None
    raw = await _pool.get(_key(prefix, user_id, n))
    if raw is None:
        CACHE_MISSES.labels(kind=prefix).inc()
        return None
    CACHE_HITS.labels(kind=prefix).inc()
    return json.loads(raw)


async def set(prefix: str, user_id: int, n: int, value: list[dict]) -> None:
    """Сохраняет результат в кэш с заданным TTL."""
    if _pool is None:
        return
    await _pool.set(_key(prefix, user_id, n), json.dumps(value),
                    ex=CACHE_TTL_SECONDS)
